proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+  (func (export "d") (result f32))
+)
+```
+Just as with module definitions, the above module type is actually an
+[abbreviation][import-abbrev] in the text format for:
+```wasm
+(module
+  (import "a" (memory 1 2))
+  (import "b" (func (param i32)))
+  (export "c" (table 1 funcref))
+  (export "d" (func (result f32)))
+)
+```
+which is the form we'll mostly use in this document.
+
+In WebAssembly there is also the separate concept of a module *instance*,
+which is the result of [instantiating][Module Instantiation] a module with
+imports. An instance type is mostly just the module type with the imports removed
+(with the only future complication being [Type Imports] used in export
+signatures). For example, the above module, when instantiated, would have
+instance type:
+```wasm
+(instance
+  (export "b" (table 1 funcref))
+  (export "c" (func (result f32)))
+)
+```
+
+Just like function types, module and instance types can either be written
+"inline" or factored out into an explicit type definition that can be reused via
+`$identifier`. (Also, just like function types, the inline form is an
+[abbreviation][typeuse-abbrev] for an implicit type definition and use.) For
+example, an instance type can be defined:
+```wasm
+(type $WasiFile (instance
+  (export "read" (func (param i32 i32 i32) (result i32)))
+  (export "write" (func (param i32 i32 i32) (result i32)))
+))
+```
+
+In many examples shown below, type definitions are needed for *both* a module
+type and the instance type produced when that module type is instantiated. In
+such cases, to avoid duplicating all the exports, a new `(exports $instanceType)`
+type expression is added which can be used to define module types. For example,
+here is the type of a module which implements the above-defined `$WasiFile`
+interface via Win32 operations:
+```wasm
+(module
+  (import "Win32" "ReadFile" (func (param i32 i32 i32 i32) (result i32)))
+  (import "Win32" "WriteFile" (func (param i32 i32 i32 i32) (result i32)))
+  (exports $WasiFile)
+)
+```
+
+Lastly, just as the current [text format conventions recommend `.wat`][WAT] as
+the extension of a file that contains a module definition, this proposal
+includes a new recommendation for text files containing a bare module or
+instance type be suffixed `.wit`. `.wit` files can be used as part of the
+toolchain ecosystem for describing a module's interface without including its
+definition. For example, this can be used to generate compatible source-language
+declarations. Like `.wat` files, `.wit` files are not consumed directly by the
+host.
+
+Module and Instance types can also be used to describe the types of imports:
+
+
+### Instance Imports
+
+Just as a function, memory, table or global can be imported by specifying a
+function, memory, table or global type, modules or instances can be imported
+by specifying a module or instance type.
+
+A single instance import is practically equivalent to individually importing
```

Comment:
That's a good question, and one I haven't fully thought through.  If we consider single-level instance imports to desugar to two-level field imports, then we have issues with the positional parameters of `instance.instantiate` (which we could address by changing `instance.instantiate` to take as immediates an array of strings, but that seems suboptimal in a different dimension).  If we consider two-level field imports to retroactively desugar to single-level instance imports, then `instance.instantiate` is happy, but we'll have to define an (obvious, but) ad hoc ordering of the desugared single-level instance imports which is what I lean towards.  WDYT?

Reply:
Yes, I agree, keeping these separate seems cleaner and will probably avoid impedance troubles.

Reply:
Just to check, when you say you agree, you mean with reframing existing two-level imports as desugaring to single-level imports?

Reply:
Oh, sorry, I misstated this. I actually meant that I'd prefer _not_ doing that. I'm not actually sure how it could even work with instance.instantiate, since the instruction type would have to be completely different for an instance vs a flattened list of members.

Reply:
I think it'd be possible to say that, in module types (which are new, so no backcompat to consider), there are only single-level imports.  Thus, if you wrote a module definition (in text or binary format) with a two-level imports, it'd have to be assigned a single-level type.  This has the obvious question of what to do with interleavings like:
```wasm
(import "a" "x" (func))
(import "b" "y" (func))
(import "a" "z" (func))
```
but I think this can be unambiguously assigned a single-level type based on order of first use of the first-string:
```wasm
(import "a" (instance (export "x" (func)) (export "z" (func))))
(import "b" (instance (export "y" (func))))
```
Since `instance.instantiate` only consults module types, there should never be any ambiguity.  I think we should be able to backwards-compatibly change the structure of the [module AST](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-module) to match, making two-level just a encoding detail.  The associated tweak to the JS API's `instantiate` would be technically observable (if you used a Proxy to observe the number of GetProps on the `importObject`: it would go from N to 1 for the first-level object) but I can't imagine the change would break anything.  The only case we'd have to think about is [WebAssembly.Module.exports()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module/exports).

As another alternative, you could imagine that all imports in a module type have an optional second string.  Then `instance.instantiate` ignores strings altogether (as it already does) and simply matches one by one.  Thus, there is a distinction between the two module types given above: the former takes 3 `func` arguments; the latter takes 2 `instance` arguments.  Presumably this non-equivalence also extends to module subtyping.  Altogether, it feels unfortunate to have two slightly different, but incompatible ways of expressing the same thing, making me wonder whether it was better to remove single-level imports altogether (with the consequence being that instance imports would rarely be used with the consequence that there'd be a lot of passing around fields individually instead of in aggregate).  I think having a single canonical form in the module type would be preferable.

Reply:
I agree that including instances as aggregates is relevant. For one, it's not clear what instantiate would return otherwise -- at least if we want to avoid index flattening, which I now strongly think we do (see my other comment).

So I'd rather go for the second alternative, where strings are ignored by an instantiate instruction. FWIW, there is more leeway for a dedicated syntax, but it might not be worth treating that differently.

Reply:
It's a bit simpler to treat the two kinds of imports separately, but what about the problem of having two slightly different, and incompatible ways to express the same thing?  It seems like this will cause unnecessary friction, given that is a problem that I think we are well able to solve.

Reply:
I don't know if there would be friction. Is a record/struct just a "slightly different way" of passing ints? :)

Reply:
It would be if we passed around ints by saying "the record X containing int Y" ;)

Reply:
Thinking about this again to try to wrap up this discussion, I realized a reasonable compromise might be, as with the ordered/unordered issue, to say that module types have both single-level and two-level imports, but that they are subtypes of each other.  I put some verbiage to this effect in the last commit.  @rossberg wdyt?

Reply:
Hm, I don't know, this design is a complication and it could have undesirable consequences, for example, it could make subtype checks more expensive algorithmically. It also is rather arbitrary that one could flatten out one level but not more. I'd prefer staying conservative and not support such interconvertibility, until there is a stronger case that it's needed.

Reply:
Here is another problem. With type imports, you could try to write something like this:
```
(type $t (import "a" "t"))
(type $u (import "b" "u") (sub $t))
(type $v (import "a" "v") (sub $u))
```
Under the current semantics, that would be fine. however, if you want 2-level imports to be interconvertible with instances, then this must not be allowed (mutually recursive instance imports!). So we would need to impose a restriction on the dependencies of imports (both types and instances).

If you want to treat 2-level imports as equivalent to one instance, then I think the cleanest way would be to say to redefine the current syntax as syntactic sugar for a respective collection of instance imports (which implies that it's just definition site, and they are not interconvertible in the instantiate construct). Then examples like the above would be syntactically illegal, though it's not clear to me how exactly to express or check that efficiently.

Reply:
Although I didn't go into detail, I had been imagining that the subtyping worked N deep, not just one level.  But that's a good counter-case for why it can't work; thanks!

So FWIW, when you say "the cleanest way would be to say to redefine the current syntax as syntactic sugar for a respective collection of instance imports", that's what I thought I was proposing [above](https://github.com/WebAssembly/module-types/pull/3#discussion_r420964856).  Is what you're saying different in a way that I'm missing or are you just warming up to it now?

Reply:
Yes, I think it's what you proposed earlier, but I wasn't sure. But we'd still need to find a way to rule out the example above. At first I wasn't sure how to do that, but I suppose it simply boils down to the question of how the desugaring would (re)order things.

There are at least two possibilities: (1) a desugared instance import is placed where its _first_ sugared import was, or (2) it is placed where its _last_ sugared import was. Either would invalidate the example above (assuming that validation requires imports to be ordered), but the choice will affect the validity of other examples, such as:
```
(type $t (import "a" "t"))
(type $u (import "b" "u") (sub $t))
(type $v (import "a" "v"))
```
vs
```
(type $t (import "a" "t"))
(type $u (import "b" "u"))
(type $v (import "a" "v") (sub $u))
```
Option 1 seems slightly more intuitive to me, but only by a margin.

Previously I though that another problem would be conflicts, i.e., what to do when you have both sugared and desugared forms for the same module name. But I realised that that doesn't actually matter: we can just desugar the nested ones independently, creating two flat imports for the same name. No problem, we already allow multiple imports of the same name, and it would have the right semantics. :)

If we go the desugaring route, then it would of course be natural to generalise the shorthand to N levels. Probably not worth introducing that into the binary format, but definitely would make sense for the text format.

Reply:
Great, thanks, that all makes sense.  I'll update accordingly.

Reply:
Ok, done.  (Supporting the explicit `alias` design choice: I realized that recasting multi-level imports as sugar for single-level imports relies on the ability of `alias` to arbitrarily order instance exports into the containing index spaces.)

Reply:
Looks good!